### PR TITLE
Support Python 3.14

### LINKED ISSUE #2070: support Python 3.14
We do not yet support Python version 3.14, which was released a few months ago. 

This would require adding it to the list in [setup.py](https://github.com/PlasmaControl/DESC/blob/master/setup.py) and making sure tests are run with it. 

We should also document which Python versions are supported, possibly with a badge or at least in the installation instructions.

### PR DESCRIPTION
**Pull request by dpanici**
Resolves #2070

Nothing major changed. Main things are
- bumps upper bound for scipy to `1.17.0` , matplotlib to `3.10.8` for numpy to `2.4.2` as those are the newest versions, which also have Python 3.14 support (though note that pip never complained when I installed with an upper bound for numpy which did not support 3.14, which was annoying...) 
  - when I installed cleanly with these bounds, the numpy got set to `2.3.5` so that seems to be the cap based on our other dependencies
- scipy 1.16.1+ deprecates `disp` for `l-bfgs-b` which was unused anyways, so had to add a check to pop that from options in that case
- I ran all of our tests on stellar with a fresh env following the CPU installation and all passed. 
- Adds 3.14 into the unit test rotation
- Move regression tests to 3.11 to prep for eventual 3.10 deprecation at end of year
- I have not updated any of the GPU installations on clusters yet as don't want to spend the time repeatedly installing and verifying (and I have no GPU locally), so if anyone happens to fresh install on Python 3.14 and verify, feel free, otherwise I will just leave them as is.

However
- `matplotlib<3.10.5` is broken with Python 3.14 due to some lowlevel changes in cPython https://github.com/matplotlib/matplotlib/issues/29185
- And pip will just allow installation of 3.9.2 with Python 3.14, which means that there are broken package combinations with Python 3.14
- And we will have to regen our test images to use matplotlib 3.10.5+ and use that version in our CI to allow Python 3.14 to be included in the tests

So a possible way forward I guess is to just
- [x] Regen our test images to use matplotlib 3.10.8
- [x] use matplotlib 3.10.8 in our CI
- [x] Accept that someone out there who installs DESC on Python 3.14 but uses an old matplotlib version will see it is broken, and hope they will understand that they need to use a newer matplotlib version to fix the changes

### CODE REVIEW

**Comment by dpanici** [comment]
Not sure why failing as it passes on my stellar install. My pip list is

```
Package                       Version
----------------------------- --------------------
alabaster                     1.0.0
aniso8601                     10.0.1
annotated-types               0.7.0
ansi2html                     1.9.2
arrow                         1.3.0
asttokens                     3.0.0
attrs                         25.3.0
babel                         2.17.0
beautifulsoup4                4.13.4
black                         24.10.0
bleach                        6.2.0
blinker                       1.9.0
build                         1.2.2.post1
certifi                       2025.7.14
cfgv                          3.4.0
cftime                        1.6.4.post1
charset-normalizer            3.4.2
choreographer                 1.1.1
click                         8.2.1
colorama                      0.4.6
comm                          0.2.3
contourpy                     1.3.3
coverage                      7.10.1
cycler                        0.12.1
debugpy                       1.8.15
decorator                     5.2.1
defusedxml                    0.7.1
desc-opt                      0.16.0+32.g0e3795667
diffrax                       0.7.0
distlib                       0.4.0
docutils                      0.21.2
equinox                       0.12.2
eradicate                     2.3.0
executing                     2.2.0
f90nml                        1.4.5
fastjsonschema                2.21.1
filelock                      3.18.0
flake8                        7.3.0
flake8-docstrings             1.7.0
flake8-eradicate              1.5.0
flake8-isort                  6.1.1
Flask                         3.1.1
Flask-RESTful                 0.3.10
fonttools                     4.59.0
h5py                          3.14.0
identify                      2.6.12
idna                          3.10
imageio                       2.37.0
imagesize                     1.4.1
iniconfig                     2.1.0
interpax                      0.3.9
ipykernel                     6.30.0
ipython                       9.4.0
ipython_pygments_lexers       1.1.1
isort                         5.13.2
itsdangerous                  2.2.0
jax                           0.6.2
jax-cuda12-pjrt               0.6.2
jax-cuda12-plugin             0.6.2
jax-finufft                   1.2.0
jaxlib                        0.6.2
jaxopt                        0.8.5
jaxtyping                     0.3.2
jedi                          0.19.2
Jinja2                        3.1.6
jsonschema                    4.25.0
jsonschema-specifications     2025.4.1
jupyter_client                8.6.3
jupyter_core                  5.8.1
jupyterlab_pygments           0.3.0
kaleido                       1.1.0
kiwisolver                    1.4.8
lazy_loader                   0.4
lineax                        0.0.8
logistro                      1.1.0
MarkupSafe                    3.0.2
matplotlib                    3.10.3
matplotlib-inline             0.1.7
mccabe                        0.7.0
memory-profiler               0.61.0
mistune                       3.1.3
ml_dtypes                     0.5.3
mpmath                        1.3.0
mypy_extensions               1.1.0
narwhals                      2.0.1
nbclient                      0.10.2
nbconvert                     7.16.6
nbformat                      5.10.4
nbmake                        1.5.5
nbsphinx                      0.9.7
nest-asyncio                  1.6.0
netCDF4                       1.7.2
networkx                      3.5
nodeenv                       1.9.1
numpy                         2.2.6
nvgpu                         0.10.0
nvidia-cublas-cu12            12.9.1.4
nvidia-cuda-cupti-cu12        12.9.79
nvidia-cuda-nvcc-cu12         12.9.86
nvidia-cuda-nvrtc-cu12        12.9.86
nvidia-cuda-runtime-cu12      12.9.79
nvidia-cudnn-cu12             9.11.0.98
nvidia-cufft-cu12             11.4.1.4
nvidia-cusolver-cu12          11.7.5.82
nvidia-cusparse-cu12          12.5.10.65
nvidia-ml-py                  12.575.51
nvidia-nccl-cu12              2.27.6
nvidia-nvjitlink-cu12         12.9.86
nvidia-nvshmem-cu12           3.3.9
opt_einsum                    3.4.0
optimistix                    0.0.10
orjson                        3.11.3
orthax                        0.2.4
packaging                     25.0
pandas                        2.3.1
pandocfilters                 1.5.1
parso                         0.8.4
pathspec                      0.12.1
pexpect                       4.9.0
pillow                        11.3.0
pip                           25.1
platformdirs                  4.3.8
plotly                        6.2.0
pluggy                        1.6.0
pre_commit                    4.2.0
prompt_toolkit                3.0.51
psutil                        7.0.0
ptyprocess                    0.7.0
pure_eval                     0.2.3
py-cpuinfo                    9.0.0
pycodestyle                   2.14.0
pydantic                      2.12.5
pydantic_core                 2.41.5
pydocstyle                    6.3.0
pyflakes                      3.4.0
Pygments                      2.19.2
pylatexenc                    2.10
pynvml                        12.0.0
pyparsing                     3.2.3
pyproject_hooks               1.2.0
pytest                        8.4.1
pytest-benchmark              5.1.0
pytest-cov                    6.2.1
pytest-monitor                1.6.6
pytest-mpl                    0.16.1
pytest-split                  0.10.0
pytest-timeout                2.4.0
python-dateutil               2.9.0.post0
pytz                          2025.2
PyYAML                        6.0.2
pyzmq                         27.0.0
qicna                         0.3.4
qsc                           0.1.3
quadax                        0.2.8
referencing                   0.36.2
requests                      2.32.4
rpds-py                       0.26.0
scikit-image                  0.25.2
scipy                         1.15.3
setuptools                    78.1.1
shapely                       2.1.1
shtab                         1.7.2
simplejson                    3.20.2
six                           1.17.0
snowballstemmer               3.0.1
soupsieve                     2.7
Sphinx                        8.1.3
sphinx-argparse               0.5.2
sphinx-copybutton             0.5.2
sphinx_design                 0.6.1
sphinx-github-style           1.2.2
sphinx-rtd-theme              3.0.2
sphinxcontrib-applehelp       2.0.0
sphinxcontrib-devhelp         2.0.0
sphinxcontrib-htmlhelp        2.1.0
sphinxcontrib-jquery          4.1
sphinxcontrib-jsmath          1.0.1
sphinxcontrib-qthelp          2.0.0
sphinxcontrib-serializinghtml 2.0.0
stack-data                    0.6.3
sympy                         1.14.0
tabulate                      0.9.0
termcolor                     3.1.0
tifffile                      2025.6.11
tinycss2                      1.4.0
tldr                          3.4.3
tokenize_rt                   6.2.0
tornado                       6.5.1
traitlets                     5.14.3
types-python-dateutil         2.9.0.20250708
typing_extensions             4.14.1
typing-inspection             0.4.2
tzdata                        2025.2
urllib3                       2.5.0
virtualenv                    20.32.0
wadler_lindig                 0.1.7
wcwidth                       0.2.13
webencodings                  0.5.1
Werkzeug                      3.1.3
wheel                         0.45.1
```

**Comment by dpanici** [comment]
yea I am not sure either. I tried to google how to deal with this and am just finding no helpful info, but did find an interesting article on version capping lol which argues the opposite of what I feel https://iscinumpy.dev/post/bound-version-constraints/

File: .github/workflows/unit_tests.yml [hunk 1]
```diff
@@ -29,7 +29,7 @@ jobs:
                  {group: 5, python_version: '3.12'},
                  {group: 6, python_version: '3.12'},
                  {group: 7, python_version: '3.13'},
-                 {group: 8, python_version: '3.13'}]
+                 {group: 8, python_version: '3.14'}]
```
Comment (YigitElma) [inline .github/workflows/unit_tests.yml:32]: Nitpick: I would remove one of the 3.10 ones, but really doesn't matter.

**Review by YigitElma — DISMISSED** [review]

**Comment by YigitElma** [comment]
I think the infinite values in master data are due to the grid nodes on the magnetic axis. We compute equilibrium quantities on axis (to make sure the limits are correctly implemented) but some of these quantities are not well-defined and cause the issue. These are the problematic values:
```
e^theta for desc.equilibrium.equilibrium.Equilibrium is not finite.
e^vartheta for desc.equilibrium.equilibrium.Equilibrium is not finite.
e^theta_PEST for desc.equilibrium.equilibrium.Equilibrium is not finite.
J^theta_PEST for desc.equilibrium.equilibrium.Equilibrium is not finite.
J^vartheta for desc.equilibrium.equilibrium.Equilibrium is not finite.
J^theta for desc.equilibrium.equilibrium.Equilibrium is not finite.
grad(alpha) (periodic) for desc.equilibrium.equilibrium.Equilibrium is not finite.
grad(alpha) for desc.equilibrium.equilibrium.Equilibrium is not finite.
g^aa for desc.equilibrium.equilibrium.Equilibrium is not finite.
g^ra for desc.equilibrium.equilibrium.Equilibrium is not finite.
g^tt for desc.equilibrium.equilibrium.Equilibrium is not finite.
e^theta_r for desc.equilibrium.equilibrium.Equilibrium is not finite.
e^helical for desc.equilibrium.equilibrium.Equilibrium is not finite.
e^theta_rr for desc.equilibrium.equilibrium.Equilibrium is not finite.
|e^helical| for desc.equilibrium.equilibrium.Equilibrium is not finite.
g^rt for desc.equilibrium.equilibrium.Equilibrium is not finite.
g^tt_r for desc.equilibrium.equilibrium.Equilibrium is not finite.
|grad(theta)| for desc.equilibrium.equilibrium.Equilibrium is not finite.
g^rv for desc.equilibrium.equilibrium.Equilibrium is not finite.
```

File: tests/test_compute_everything.py [hunk 2]
```diff
@@ -43,6 +43,11 @@ def _compare_against_master(
                 mean = 1.0
             else:
                 mean = np.nanmean(np.atleast_1d(np.abs(master_data[p][name])))
+                if not np.isfinite(mean):
+                    # since eq grid has nodes on axis, some quantities are infinite
+                    # so just use 1.0 as a scale for the tolerance
+                    print(f"mean(|master data|) of {name} for {p} is not finite.")
+                    mean = 1.0
```
Comment (dpanici) [inline tests/test_compute_everything.py:50]: we could also instead do
```python
this_data=np.atleast_1d(np.abs(master_data[p][name]))
finite_data = this_data[np.where(np.isfinite(this_data))]
mean = np.nanmean(finite_data)
```

maybe nanmean not necessary here anymore, but this would keep the mean part. Maybe not needed though, if what you have works

Comment (YigitElma) [reply tests/test_compute_everything.py:50]: `atol=1e-8 * mean + 1e-9` so I guess anything is fine

File: tests/test_examples.py [hunk 3]
```diff
@@ -169,9 +169,9 @@ def test_solve_bounds():
         eq.change_resolution(L=eq.M, L_grid=eq.M_grid)
     eq.p_l *= 2
 
-    # target force balance residuals with |F| <= 1e3 N
+    # target force balance residuals with |F| <= 3e3 N
     obj = ObjectiveFunction(
-        ForceBalance(normalize=False, normalize_target=False, bounds=(-1e3, 1e3), eq=eq)
+        ForceBalance(normalize=False, normalize_target=False, bounds=(-3e3, 3e3), eq=eq)
```
Comment (YigitElma) [inline tests/test_examples.py:174]: On my laptop, `1e3` passes, `2e3` fails, `3e3` and higher pass. I guess there are some edge cases for bounds

**Review by YigitElma — DISMISSED** [review]
Someone else should also review

File: tests/test_examples.py (hunk 3 above)
Comment (dpanici) [reply tests/test_examples.py:174]: For me locally, 1e3 2e3 and 3e3 all pass this test just fine. Seems something very sensitive to the local setup is going on?

Comment (YigitElma) [reply tests/test_examples.py:174]: Did you use the new version of numpy and scipy? I tested this in a new environment with exactly the same as this PR

**Review by dpanici — COMMENTED** [review]
I can't approve as I made the main changes but I'd approve, this looks good to me and hopefully fixes dependabot now

File: tests/test_examples.py (hunk 3 above)
Comment (dpanici) [reply tests/test_examples.py:174]: oh no I did not use the new versions

Comment (dpanici) [reply tests/test_examples.py:174]: on stellar with Python 3.14 and scipy 1.17 and numpy 2.3.5 all three seem to pass for me. I did not check the rest of the environment for the PR though

Comment (YigitElma) [reply tests/test_examples.py:174]: maybe numpy 2.4.2 has some changes?

File: .python-version
Comment (f0uriest) [inline .python-version:1]: can we get away with just `3.13`? or do we need to specify the patch version? in which case should probably at least use the latest

Comment (dpanici) [reply .python-version:1]: Probably we cna just do 3.13

**Review by YigitElma — DISMISSED** [review]

**Review by YigitElma — DISMISSED** [review]

**Review by YigitElma — DISMISSED** [review]
Note: Rory tested the newer versions of Jax locally, and it works except the `jax-finufft` tests

File: requirements.txt [hunk 4]
```diff
@@ -8,14 +8,14 @@ jax-finufft >= 1.1.0
...
+numpy >= 1.20.0, <= 2.4.2
 nvgpu <= 0.10.0
 orthax <= 0.2.7
 plotly >= 5.16, <= 6.5.2
 psutil <= 7.2.2
 pylatexenc >= 2.0, <= 2.10
 quadax >= 0.2.2, <= 0.2.11
 scikit-image <= 0.26.0
-scipy >= 1.7.0, <= 1.15.3
+scipy >= 1.7.0, <= 1.17.0
```
Comment (YigitElma) [inline requirements.txt]:
```suggestion
scipy >= 1.7.0, <= 1.17.1
```

**Review by YigitElma — APPROVED** [review]

**Review by ddudt — APPROVED** [review]

File: tests/test_compute_everything.py (hunk 2 above)
Comment (unalmis) [reply tests/test_compute_everything.py:50]: This was already fixed on #1919 in september
